backend/src/main/kotlin/org/loculus/backend/model/SubmitModel.kt

```diff
@@ -400,6 +361,58 @@ class SubmitModel(
         }
     }
 
+    private fun validateSubmissionIdSetsForFiles(metadataKeysSet: Set<SubmissionId>, filesKeysSet: Set<SubmissionId>) {
+        val filesKeysNotInMetadata = filesKeysSet.subtract(metadataKeysSet)
+        if (filesKeysNotInMetadata.isNotEmpty()) {
+            throw UnprocessableEntityException(
+                "Sequence file contains ${filesKeysNotInMetadata.size} submissionIds that are not present " +
+                    "in the metadata file: " + filesKeysNotInMetadata.toList().joinToString(limit = 10),
+            )
+        }
+    }
+
+    private fun validateFileExistenceAndGroupOwnership(
+        submittedFiles: SubmissionIdFilesMap,
+        submissionParams: SubmissionParams,
+        uploadId: String,
+    ) {
+        val usedFileIds = submittedFiles.getAllFileIds()
+        val fileGroups = filesDatabaseService.getGroupIds(usedFileIds)
+
+        log.debug { "Validating that all submitted file IDs exist." }
+        val notExistingIds = usedFileIds.subtract(fileGroups.keys)
+        if (notExistingIds.isNotEmpty()) {
+            throw BadRequestException("The File IDs $notExistingIds do not exist.")
+        }
+
+        log.debug {
+            "Validating that submitted files belong to the group that their associated submission belongs to."
+        }
+        if (submissionParams is SubmissionParams.OriginalSubmissionParams) {
+            fileGroups.forEach {
+                if (it.value != submissionParams.groupId) {
+                    throw BadRequestException(
+                        "The File ${it.key} does not belong to group ${submissionParams.groupId}.",
```

Comment:
I changed it to not print the group the file currently belongs to to avoid "leaking" this information.

Reply:
Oh, good idea!